src/AutoRest.CSharp/Output/Models/RestClientHelpers.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using AutoRest.CSharp.Generation.Types;
+using AutoRest.CSharp.Input;
+using AutoRest.CSharp.Output.Builders;
+using AutoRest.CSharp.Output.Models.Requests;
+using AutoRest.CSharp.Output.Models.Responses;
+using AutoRest.CSharp.Output.Models.Serialization;
+using AutoRest.CSharp.Output.Models.Shared;
+using AutoRest.CSharp.Output.Models.Types;
+using AutoRest.CSharp.Utilities;
+using Azure.Core;
+using Request = AutoRest.CSharp.Output.Models.Requests.Request;
+using StatusCodes = AutoRest.CSharp.Output.Models.Responses.StatusCodes;
+
+namespace AutoRest.CSharp.Output.Models
+{
+    internal static class RestClientHelpers
+    {
+        private static IEnumerable<Parameter> GetRequiredParameters(Parameter[] parameters)
+        {
+            List<Parameter> requiredParameters = new List<Parameter>();
+            foreach (var parameter in parameters)
+            {
+                if (parameter.DefaultValue == null)
+                {
+                    requiredParameters.Add(parameter);
+                }
+            }
+
+            return requiredParameters;
+        }
+
+        private static IEnumerable<Parameter> GetOptionalParameters(Parameter[] parameters, bool includeAPIVersion = false)
+        {
+            List<Parameter> optionalParameters = new List<Parameter>();
+            foreach (var parameter in parameters)
+            {
+                if (parameter.DefaultValue != null && (includeAPIVersion || !parameter.IsApiVersionParameter))
+                {
+                    optionalParameters.Add(parameter);
+                }
+            }
+
+            return optionalParameters;
+        }
+
+        public static IReadOnlyCollection<Parameter> GetConstructorParameters(Parameter[] parameters, CSharpType credentialType, bool includeProtocolOptions = false, bool includeAPIVersion = false)
+        {
+            List<Parameter> constructorParameters = new List<Parameter>();
+
+            constructorParameters.AddRange(GetRequiredParameters(parameters));
+
+            var credentialParam = new Parameter(
+                "credential",
+                "A credential used to authenticate to an Azure Service.",
+                credentialType,
+                null,
+                true);
+            constructorParameters.Add(credentialParam);
+
+            if (includeProtocolOptions)
```

Comment:
Should this logic go onto the LLC client? Don't think DataPlane cares about it

Reply:
We could, but it was a bit of pain, since we want the parameter after all required before any optional, and we don't have a clear idea where that split point is outside of the function.

I'd have to split GetConstructorParameters into GetConstructorParametersRequired and GetConstructorParametersOptional or something, and have LLC glue it in between.

Thoughts?

Reply:
Fine for now.